File: hu/codes/dart/chapter_backtracking/permutations_ii.dart
/**
 * File: permutations_ii.dart
 * Created Time: 2023-08-10
 * Author: liuyuxin (gvenusleo@gmail.com)
 */

/* Visszalépéses algoritmus: Permutációk II */
void backtrack(
  List<int> state,
  List<int> choices,
  List<bool> selected,
  List<List<int>> res,
) {
  // Ha az állapot hossza egyenlő az elemek számával, jegyezze fel a megoldást
  if (state.length == choices.length) {
    res.add(List.from(state));
    return;
  }
  // Összes választási lehetőség bejárása
  Set<int> duplicated = {};
  for (int i = 0; i < choices.length; i++) {
    int choice = choices[i];
    // Metszés: ne engedélyezd az elemek ismételt kiválasztását és az egyenlő értékű elemek ismételt kiválasztását
    if (!selected[i] && !duplicated.contains(choice)) {
      // Kísérlet: válaszd ki, frissítsd az állapotot
      duplicated.add(choice); // A kiválasztott elem értékének rögzítése
      selected[i] = true;
      state.add(choice);
      // Lépj a következő kiválasztási körre
      backtrack(state, choices, selected, res);
      // Visszalépés: vonja vissza a választást, állítsa vissza az előző állapotot
      selected[i] = false;
      state.removeLast();
    }
  }
}

/* Permutációk II */
List<List<int>> permutationsII(List<int> nums) {
  List<List<int>> res = [];
  backtrack([], nums, List.filled(nums.length, false), res);
  return res;
}

/* Főprogram */
void main() {
  List<int> nums = [1, 2, 2];

  List<List<int>> res = permutationsII(nums);

  print("Input array nums = $nums");
  print("All permutations res = $res");
}

File: hu/codes/javascript/chapter_backtracking/permutations_ii.js
/**
 * File: permutations_ii.js
 * Created Time: 2023-05-13
 * Author: Justin (xiefahit@gmail.com)
 */

/* Visszalépéses algoritmus: Permutációk II */
function backtrack(state, choices, selected, res) {
    // Ha az állapot hossza egyenlő az elemek számával, rögzítjük a megoldást
    if (state.length === choices.length) {
        res.push([...state]);
        return;
    }
    // Bejárjuk az összes választási lehetőséget
    const duplicated = new Set();
    choices.forEach((choice, i) => {
        // Metszés: nem engedünk meg elemismétlést, és nem engedünk meg egyenlő elemek ismételt kiválasztását
        if (!selected[i] && !duplicated.has(choice)) {
            // Kísérlet: választás megtétele, állapot frissítése
            duplicated.add(choice); // Rögzítjük a kiválasztott elem értékét
            selected[i] = true;
            state.push(choice);
            // Folytatjuk a következő kiválasztási körrel
            backtrack(state, choices, selected, res);
            // Visszalépés: a választás visszavonása, visszaállítás az előző állapotra
            selected[i] = false;
            state.pop();
        }
    });
}

/* Permutációk II */
function permutationsII(nums) {
    const res = [];
    backtrack([], nums, Array(nums.length).fill(false), res);
    return res;
}

// Tesztkód
const nums = [1, 2, 2];
const res = permutationsII(nums);

console.log(`Input array nums = ${JSON.stringify(nums)}`);
console.log(`All permutations res = ${JSON.stringify(res)}`);
